sway/input/text_input.c

```diff
@@ -255,6 +268,220 @@ static void relay_handle_text_input(struct wl_listener *listener,
 	sway_text_input_create(relay, wlr_text_input);
 }
 
+static void input_popup_update(struct sway_input_popup *popup) {
+	struct sway_text_input *text_input =
+		relay_get_focused_text_input(popup->relay);
+
+	if (text_input == NULL || text_input->input->focused_surface == NULL) {
+		return;
+	}
+
+	if (popup->scene_tree != NULL) {
+		wlr_scene_node_destroy(&popup->scene_tree->node);
+		popup->scene_tree = NULL;
+	}
+	if (popup->desc.relative != NULL) {
+		wlr_scene_node_destroy(popup->desc.relative);
+		popup->desc.relative = NULL;
+	}
+	popup->desc.view = NULL;
+
+	if (!popup->popup_surface->surface->mapped) {
+		return;
+	}
+
+	wlr_scene_node_destroy(&popup->scene_tree->node);
+	wlr_scene_node_destroy(popup->desc.relative);
+	popup->scene_tree = NULL;
+	popup->desc.view = NULL;
+
+	bool cursor_rect = text_input->input->current.features
+		& WLR_TEXT_INPUT_V3_FEATURE_CURSOR_RECTANGLE;
+	struct wlr_surface *focused_surface = text_input->input->focused_surface;
+	struct wlr_box cursor_area = text_input->input->current.cursor_rectangle;
+
+	struct wlr_box output_box;
+	struct wlr_box parent;
+	struct wlr_layer_surface_v1 *layer_surface =
+		wlr_layer_surface_v1_try_from_wlr_surface(focused_surface);
+	struct wlr_scene_tree *relative;
+	if (layer_surface != NULL) {
+		struct sway_layer_surface *layer =
+			layer_surface->data;
+		if (layer == NULL) {
+			return;
+		}
+
+		relative = wlr_scene_tree_create(layer->scene->tree);
+		if (relative == NULL) {
+			return;
+		}
+		struct wlr_output *output = layer->layer_surface->output;
+		wlr_output_layout_get_box(root->output_layout, output, &output_box);
+		int lx, ly;
+		wlr_scene_node_coords(&layer->tree->node, &lx, &ly);
+		parent.x = lx;
+		parent.y = ly;
+		popup->scene_tree = wlr_scene_subsurface_tree_create(root->layers.popup, popup->popup_surface->surface);
+		popup->desc.view = NULL;
+	} else {
+		struct sway_view *view = view_from_wlr_surface(focused_surface);
+		relative = wlr_scene_tree_create(view->scene_tree);
+		if (relative == NULL) {
+			return;
+		}
+		int lx, ly;
+		wlr_scene_node_coords(&view->scene_tree->node, &lx, &ly);
+		struct wlr_output *output = wlr_output_layout_output_at(root->output_layout,
+			view->container->pending.content_x + view->geometry.x,
+			view->container->pending.content_y + view->geometry.y);
+		wlr_output_layout_get_box(root->output_layout, output, &output_box);
+		parent.x = lx;
+		parent.y = ly;
+
+		parent.width = view->geometry.width;
+		parent.height = view->geometry.height;
+		popup->scene_tree = wlr_scene_subsurface_tree_create(root->layers.popup, popup->popup_surface->surface);
+		popup->desc.view = view;
+	}
```

Comment:
This comment has been lost, but hasn't been addressed yet: https://github.com/swaywm/sway/pull/7226#discussion_r1487159669

(Just in case you missed it.)

Reply:
ok, done